content/blog/prototyping/index.md

```diff
@@ -0,0 +1,732 @@
+
+| Prototype        | Production          | When to switch                                                                     |
+| ---------------- | ------------------- | ---------------------------------------------------------------------------------- |
+| `String`         | `&str`              | When you need to avoid allocations or store string data with a clear lifetime      |
+| `Vec<T>`         | `&[T]`              | When the owned vector becomes too expensive to clone or you can't afford the heap  |
+| `Box<T>`         | `&T` or `&mut T`    | When `Box` becomes a bottleneck or you don't want to deal with heap allocations    | 
+| `Rc<T>`          | `&T`                | When the reference counting overhead becomes too expensive or you need mutability  | 
+| `Arc<Mutex<T>>`  | `&mut T`            | When you can guarantee exclusive access and don't need thread safety               |
+| `String`         | `Cow<'static, str>` | When you want to optimize string reuse or avoid allocations                        |
+
+These owned types sidestep most ownership and lifetime issues, but they do it by allocating memory on the heap - just like Python or JavaScript would.
+You can always refactor when you need to optimize for performance or tighten up resource usage, but chances are you won't need to.
+
+### Make use of type inference
+
+Rust is a statically, strongly typed language.
+It would be tedious to write out all the types all the time if it weren't for Rust's type inference.
+
+You can often omit (also called "elide") the types and let the compiler figure it out from the context.
+
+```rust
+let x = 42;
+let y = "hello";
+let z = vec![1, 2, 3];
+```
+
+This is a great way to get started quickly and defer the decision about types to later.
+The system scales well with more complex types, so you can use this technique even in larger projects.
+
+```rust
+let x: Vec<i32> = vec![1, 2, 3];
+let y: Vec<i32> = vec![4, 5, 6];
+
+// Rust can infer the item type of z
+let z = x.iter().chain(y.iter()).collect::<Vec<_>>();
+```
+
+Here's a more complex example which shows just how powerful Rust's type inference can be: 
+
+```rust
+use std::collections::HashMap;
+
+// Start with some nested data
+let data = vec![
+    ("fruits", vec!["apple", "banana"]),
+    ("vegetables", vec!["carrot", "potato"]),
+];
+
+// Let Rust figure out this complex transformation
+let categorized = data
+    .into_iter()
+    .flat_map(|(category, items)| {
+        items.into_iter().map(move |item| (item, category))
+    })
+    .collect::<HashMap<_, _>>();
+
+// categorized is now a HashMap<&str, &str> mapping items to their categories
+println!("What type is banana? {}", categorized.get("banana").unwrap());
+```
+
+([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fac339eecef40b69b919a1670a0a53df))
+
+It's not easy to visualize the structure of `categorized` in your head, but Rust can do it for you!
```

Comment:
Should mention turning on inlay type hints in the editor

Reply:
Done. Thanks for the idea.